policies/engineering-principles.md

```diff
@@ -0,0 +1,50 @@
+
+### The Principle of Code Purity
+
+Whenever possible prefer pure, functional code over imperative code. Pure code is free of side effects, and will return the same output given the same inputs. Imperative code is code that produces side effects, such as mutation or state changes. Imperative code is not avoidable but we should strive to keep as much of our application logic pure as possible. 
+
+
+### The Principle of Incremental Change
+
+To minimize the risk to the stability and security of the application, changes to the code should be incremental and as small as possible. Large changes to the codebase become difficult to review effectively and make the commit history less itemized. Furthermore, incremental change promotes better test coverage and makes it easier to require tests to cover the introduced changes.
+
+### The Principle of Incremental Decentralization
+
+We may sometimes ship with centralized services by default, but if we must, users should be able to opt out of any of these during onboarding, ideally easily swapping them out for decentralized alternatives. When we build features that rely upon centralized services we must build them in such a way where the experience isn't completely broken when the service is disrupted. We are trying to build a decentralized future and we must make choices that move us directionally towards those goals.
+
+### The Principle of Self-Soverign Identity
+
+We believe in the right of our users to own their own identity and data. We must go to great lengths to protect that right. Where we have features or services that might infringe upon that right we must treat those features as opt in, not opt out. When we do build features that make user data known to us we must not collect any personally identifiable information, and any data that can be tied to any chain must be collected in an anonymity protecting way. This principle must be extended to cover all third party code as well.
+
+### The Principle of Auditability
+
+All code contributions must be reviewed thoroughly for code quality, usability and adherence to these principles. Code contributions must be kept as small as possible (see Principle of Incremental Change) to promote auditability. Avoid long running feature branches because they are not auditable, and pose a risk to the stability and security of the product. We should engage in self-audit of code that we contribute, as well as actively engage in code review of others' pull requests to ensure best practices, and engineering principles are reflected in the work.
```

Comment:
This might be clearer if we frame it as a commitment to ensuring our work is auditable, rather than as a commitment to actively engage in review.

e.g. We could start by stating a commitment to transparency (should this be a value as well? :thinking: ), and/or making a connection between security and the codebase being publicly accessible.

If our codebase:
* is easy to read
* follows consistent conventions
* maintains a high standard for code usability/quality
* has a public record of why changes are made (the git/PR history)

then our codebase will be more maintainable as a direct result, and it will be easier for us to uphold our commitments to security and stability. And users can trust us less by directly auditing the code themselves, or hiring third parties to audit our codebase (especially important as part of a postmortem when things go wrong).

We've paid for third-party audits before, so this isn't even a theoretical concern. This would directly benefit us even in that sense.

This framing of auditability makes more sense as a "principle" to me. Whereas practices like code review, standards for accepting contributions, practices like long-running feature branches, those are.... well, they're details, means to an end. Of course we want tall of those things but we want to maintain a high degree of team autonomy to adjusting their development practices to best suit their own needs, and I don't think we need to commit to them at this level.

Reply:
Let me know if that makes sense, and I can try to produce a coherent draft. I tried just now but had to give up and revert to the bulleted list.

Reply:
I agree that this reads more like a list of rules than a principle that those rules could be based on. I think reframing this in terms of transparency a lot more sense.

Reply:
That said, we could also change this in a future PR.

Reply:
I took a stab at words to capture your intent here @Gudahtt , but i'm not 100% on it. Could you take a look and make suggestions? https://github.com/MetaMask/contributor-docs/pull/6/commits/0dcb4cc296adb46eca7570ef302ccc82e4b97688